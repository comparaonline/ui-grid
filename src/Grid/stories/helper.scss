.row {
  background: #eee;
  font-size: 11px;
  margin-bottom: 10px;
  line-height: 2rem;
  border: 1px solid #c6c6c6;
  margin-left: 0;
  margin-right: 0;
}

.row .column:nth-child(2) {
  background: #e1e1e1;
}
