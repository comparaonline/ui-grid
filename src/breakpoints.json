{
  "small": "0",
  "medium": "200px",
  "large": "1024px",
  "xlarge": "1170px",
  "xxlarge": "1306px"
}
