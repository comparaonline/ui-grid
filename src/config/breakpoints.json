{
  "xsmall": "0",
  "small": "480px",
  "medium": "600px",
  "large": "840px",
  "xlarge": "960px",
  "xxlarge": "1280px"
}
