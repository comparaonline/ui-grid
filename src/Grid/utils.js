import PropTypes from 'prop-types';
import breakpoints from '../breakpoints.json';

const bps = Object.keys(breakpoints);
const bpCamelCase = bp => `${bp.substr(0, 1).toUpperCase()}${bp.substr(1)}`;

export const bpPropClass = (bp, prop) => `${prop}On${bpCamelCase(bp)}`;

const getBooleanClassNames = (props, booleanProps) => booleanProps.reduce(
  (names, prop) => names.concat(props[prop] ? ` ${prop}` : ''), '');

const getBreakpointClassName = (props, bp) => (props[bp] ? ` ${bp}-${props[bp]}` : '');

const getBreakpointNumericClassName = (props, numericProps, bp) => numericProps.reduce(
  (names, prop) => {
    const propValue = props[bpPropClass(bp, prop)];
    return names.concat(propValue ? ` ${bp}-${prop}-${propValue}` : '');
  }, '');

const getBreakpointBooleanClassName = (props, booleanProps, bp) => booleanProps.reduce(
  (names, prop) => names.concat(props[bpPropClass(bp, prop)] ? ` ${bp}-${prop}` : ''), '');

const getBreakpointVisibilityClassName = (props, booleanProps, bp) => booleanProps.reduce(
  (names, prop) => names.concat(props[bpPropClass(bp, prop)] ? ` ${prop}-for-${bp}` : ''), '');

const generateBreakpointClassNames = (props, propNames, bp) => ''.concat(
  propNames.include ? getBreakpointClassName(props, bp) : '',
  propNames.numeric ? getBreakpointNumericClassName(props, propNames.numeric, bp) : '',
  propNames.boolean ? getBreakpointBooleanClassName(props, propNames.boolean, bp) : '',
  propNames.visibility ? getBreakpointVisibilityClassName(props, propNames.visibility, bp) : ''
);

/**
 * Generate string with classnames from props names and breakpoints
 *
 * @param {Object} props
 * @param {String Array} numericProps
 * @param {String Array} booleanProps
 * @param {Boolean} hasBreakpoint
 * @returns {String}
 */
export const generateClassNames = (props, propNames) => ''.concat(
  propNames.boolean ? getBooleanClassNames(props, propNames.boolean) : '',
  propNames.breakpoints ?
    bps.reduce((className, bp) => className.concat(
      generateBreakpointClassNames(props, propNames.breakpoints, bp)
    ), '') : ''
);

const getBooleanTypes = booleanProps => booleanProps.reduce(
  (result, prop) => Object.assign(result, { [prop]: PropTypes.bool }), {});

const getBreakpointType = bp => ({ [bp]: PropTypes.number });

const getBreakpointGenericType = (props, bp, type) => props.reduce(
  (result, prop) => Object.assign(result, { [bpPropClass(bp, prop)]: type }), {});

const generateBreakpointTypes = (propNames, bp) => Object.assign(
  propNames.include ? getBreakpointType(bp) : {},
  propNames.numeric ? getBreakpointGenericType(propNames.numeric, bp, PropTypes.number) : {},
  propNames.boolean ? getBreakpointGenericType(propNames.boolean, bp, PropTypes.bool) : {},
  propNames.visibility ? getBreakpointGenericType(propNames.visibility, bp, PropTypes.bool) : {}
);

/**
 * Generate object with prop types from props names and breakpoints
 *
 * @param {String Array} numericProps
 * @param {String Array} booleanProps
 * @param {Boolean} hasBreakpoint
 * @returns {Object}
 */
export const generatePropTypes = (propNames) =>
  Object.assign(
    propNames.boolean ? getBooleanTypes(propNames.boolean) : {},
    bps.reduce((result, bp) => Object.assign(
      result,
      generateBreakpointTypes(propNames.breakpoints, bp)
    ), {})
  );
