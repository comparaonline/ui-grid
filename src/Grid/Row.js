import React from 'react';
import { generateClassnames, generatePropTypes } from './utils';

export const numericProps = ['up', 'collapse', 'uncollapse'];
export const booleanProps = ['collapse', 'expanded'];

export const Row = (props) => {
  const className = generateClassnames(props, numericProps, booleanProps);
  return <div {...props} className={`row${className}`} />;
};

Row.propTypes = generatePropTypes(numericProps, booleanProps);
