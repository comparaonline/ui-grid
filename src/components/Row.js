import React from 'react';
import cx from 'classnames';
import { generateClassNames, generatePropTypes } from './utils';

export const propNames = {
  breakpoints: {
    include: false,
    numeric: ['up'],
    boolean: ['collapse', 'uncollapse'],
    visibility: ['hide', 'show']
  },
  boolean: ['collapse', 'expanded']
};

const Row = (props) => {
  const className = cx(
    `row${generateClassNames(props, propNames)}`,
    { 'flex-row': props.equalHeightColumns },
    props.className
  );

  return (
    <div className={className}>
      {props.children}
    </div>
  );
};

Row.propTypes = generatePropTypes(propNames);

export default Row;
