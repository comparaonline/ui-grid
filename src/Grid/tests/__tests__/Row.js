import React from 'react';
import { shallow } from 'enzyme';
import Row, { propNames } from '../../Row';
import { getBreakpointPropName } from '../../utils';
import breakpoints from '../../../breakpoints.json';
import { expectClass } from '../helper';

describe('Row', () => {
  it('renders a default row', () => {
    const row = shallow(<Row />);
    expectClass(row, 'row');
    expect(row.is('div')).toBe(true);
  });

  it('renders a row with numeric props', () => {
    Object.keys(breakpoints).forEach(bp => {
      propNames.breakpoints.numeric.forEach(prop => {
        const row = shallow(<Row {...{ [getBreakpointPropName(bp, prop)]: 1 }} />);
        expectClass(row, `${bp}-${prop}-1`);
      });
    });
  });

  it('renders a row with boolean props', () => {
    propNames.boolean.forEach(prop => {
      const row = shallow(<Row {...{ [prop]: true }} />);
      expectClass(row, prop);
    });
  });

  it('renders a row with breakpoints boolean props', () => {
    Object.keys(breakpoints).forEach(bp => {
      propNames.breakpoints.boolean.forEach(prop => {
        const row = shallow(<Row {...{ [getBreakpointPropName(bp, prop)]: true }} />);
        expectClass(row, `${bp}-${prop}`);
      });
    });
  });

  it('renders a row with visibility props', () => {
    Object.keys(breakpoints).forEach(bp => {
      propNames.breakpoints.visibility.forEach(prop => {
        const row = shallow(<Row {...{ [getBreakpointPropName(bp, prop)]: true }} />);
        expectClass(row, `${prop}-for-${bp}`);
      });
    });
  });

  it('renders a row with equal height columns', () => {
    const row = shallow(<Row equalHeightColumns />);
    expectClass(row, 'flex-row');
  });

  describe('when an additional className is pass as prop', () => {
    const className = 'test-class';
    const wrapper = shallow(<Row className={className} />);

    it('add this className', () => {
      expect(wrapper.is(`.${className}`)).toBeTruthy();
    });
  });
});
