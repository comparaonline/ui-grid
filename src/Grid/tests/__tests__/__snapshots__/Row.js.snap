// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row renders a default row 1`] = `
<div
  className="row"
/>
`;

exports[`Row renders a row with boolean props 1`] = `
<div
  className="row collapse"
/>
`;

exports[`Row renders a row with boolean props 2`] = `
<div
  className="row expanded"
/>
`;

exports[`Row renders a row with breakpoints boolean props 1`] = `
<div
  className="row small-collapse"
/>
`;

exports[`Row renders a row with breakpoints boolean props 2`] = `
<div
  className="row small-uncollapse"
/>
`;

exports[`Row renders a row with breakpoints boolean props 3`] = `
<div
  className="row medium-collapse"
/>
`;

exports[`Row renders a row with breakpoints boolean props 4`] = `
<div
  className="row medium-uncollapse"
/>
`;

exports[`Row renders a row with breakpoints boolean props 5`] = `
<div
  className="row large-collapse"
/>
`;

exports[`Row renders a row with breakpoints boolean props 6`] = `
<div
  className="row large-uncollapse"
/>
`;

exports[`Row renders a row with breakpoints boolean props 7`] = `
<div
  className="row xlarge-collapse"
/>
`;

exports[`Row renders a row with breakpoints boolean props 8`] = `
<div
  className="row xlarge-uncollapse"
/>
`;

exports[`Row renders a row with numeric props 1`] = `
<div
  className="row small-up-1"
/>
`;

exports[`Row renders a row with numeric props 2`] = `
<div
  className="row medium-up-1"
/>
`;

exports[`Row renders a row with numeric props 3`] = `
<div
  className="row large-up-1"
/>
`;

exports[`Row renders a row with numeric props 4`] = `
<div
  className="row xlarge-up-1"
/>
`;
